### PR TITLE
bpo-27584: New addition of vSockets to the python socket module

### PR DESCRIPTION
**Pull request by caavery**
Support for AF_VSOCK on Linux only

https://bugs.python.org/issue27584

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by caavery** [comment]
The patch is also in Issue number 27584 along with a README file that explains how to setup the environment for tests.

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -1227,6 +1227,14 @@ makesockaddr(SOCKET_T sockfd, struct sockaddr *addr, size_t addrlen, int proto)
        }
 #endif /* AF_NETLINK */
 
+#if defined(AF_VSOCK)
+       case AF_VSOCK:
+       {
+           struct sockaddr_vm *a = (struct sockaddr_vm *) addr;
+           return Py_BuildValue("II", a->svm_cid, a->svm_port);
```
Comment (kushaldas) [inline Modules/socketmodule.c:1232]: This is the line failing in the Travis. https://travis-ci.org/python/cpython/jobs/248412030#L2063

File: aclocal.m4 [hunk 2]
```diff
@@ -1,6 +1,6 @@
-# generated automatically by aclocal 1.15 -*- Autoconf -*-
+# generated automatically by aclocal 1.13.4 -*- Autoconf -*-
```
Comment (kushaldas) [inline aclocal.m4]: Why are we using an old version here?

**Review by kushaldas — CHANGES_REQUESTED** [review]

File: Modules/socketmodule.c (hunk 1 above)
Comment (caavery) [reply Modules/socketmodule.c:1232]: You don't have the proper headers. 
checking for linux/vm_sockets.h... no
 
So it looks like my checking for HAVE_LINUX_VM_SOCKETS_H is faulty.

File: aclocal.m4 (hunk 2 above)
Comment (caavery) [reply aclocal.m4]: I must have an older version installed. automake-1.13.4-3.el7.noarch I can upgrade that.

So do I make the corrections and ask for another pull?

Comment (kushaldas) [reply aclocal.m4]: Just update this PR.

File: Modules/socketmodule.c (hunk 1 above)
Comment (kushaldas) [reply Modules/socketmodule.c:1232]: Okay, so please update that, and push in your branch, the bots will pick up the changes and will test.

File: Lib/test/test_socket.py [hunk 3]
```diff
@@ -75,12 +94,20 @@ def _have_socket_alg():
         s.close()
     return True
 
+def _have_socket_vsock():
+    """Check whether AF_VSOCK sockets are supported on this host."""
+    if (getCid() == None):
```
Comment (berkerpeksag) [inline Lib/test/test_socket.py]: I'd use a shorter version:

```py
return getCid() is not None
```

File: Lib/test/test_socket.py [hunk 4]
```diff
@@ -191,7 +218,6 @@ def setUp(self):
         except OSError:
             self.skipTest('unable to bind RDS socket')
 
-
```
Comment (berkerpeksag) [inline Lib/test/test_socket.py]: Please don't make unrelated cosmetic changes.

File: Lib/test/test_socket.py [hunk 5]
```diff
@@ -5627,6 +5734,10 @@ def test_main():
     tests.extend([BasicRDSTest, RDSTest])
     tests.append(LinuxKernelCryptoAPI)
     tests.extend([
+        BasicVSOCKTest,
+        ThreadedVSOCKSocketStreamTest
```
Comment (berkerpeksag) [inline Lib/test/test_socket.py]: Style nit: Please add trailing comma.

File: Lib/test/test_socket.py [hunk 6]
```diff
@@ -44,6 +46,23 @@
 except ImportError:
     _socket = None
 
+def getCid():
+    import struct
+    import fcntl
```
Comment (berkerpeksag) [inline Lib/test/test_socket.py]: You may want to check if fcntl is available on the platform.

File: Doc/library/socket.rst [hunk 7]
```diff
@@ -149,6 +149,14 @@ created.  Socket addresses are represented as follows:
 
   .. versionadded:: 3.6
 
+- :const:`AF_VSOCK` allows communication between virtual machines and
+   their hosts. The sockets are represented as a (CID, port) tuple
```
Comment (berkerpeksag) [inline Doc/library/socket.rst]: Indentation is off here:

```rst
- :const:`AF_VSOCK` allows [...]
  their hosts. [...]

  Availability: Linux >= 4.8. QEMU >= 2.8.
```

File: Doc/library/socket.rst [hunk 8]
```diff
@@ -149,6 +149,14 @@ created.  Socket addresses are represented as follows:
...
+   where the context ID or CID and port are integers.
+
+  Availability: Linux >= 4.8. QEMU >= 2.8.
```
Comment (berkerpeksag) [inline Doc/library/socket.rst]: Do we need to mention QEMU here? I'd say just document kernel version.

File: Doc/library/socket.rst (hunk 7 above)
Comment (berkerpeksag) [inline Doc/library/socket.rst]: Style nit: (CID, port) -> \`\`(CID, port)\`\`

File: Lib/test/test_socket.py [hunk 9]
```diff
@@ -44,6 +46,23 @@
 except ImportError:
     _socket = None
 
+def getCid():
```
Comment (berkerpeksag) [inline Lib/test/test_socket.py]: Style nit: Can we use an under_score name here?

File: Lib/test/test_socket.py [hunk 10]
```diff
@@ -375,6 +401,40 @@ def clientTearDown(self):
         self.cli = None
         ThreadableTest.clientTearDown(self)
 
+@unittest.skipUnless(HAVE_SOCKET_VSOCK,
+          'VSOCK sockets required for this test.')
+@unittest.skipUnless(getCid() != 2,
+          "This test can only be run on a virtual guest.")
+class ThreadedVSOCKSocketStreamTest(unittest.TestCase, ThreadableTest):
+
+    def __init__(self, methodName = 'runTest'):
```
Comment (berkerpeksag) [inline Lib/test/test_socket.py]: Style nit: No need to add spaces around `=`.

```py
methodName= 'runTest'
```

Same for line 411:

```py
methodName=methodName
```

**Review by berkerpeksag — COMMENTED** [review]
Thank you for the PR. Could you please add your name to `Misc/ACKS`?

File: Lib/test/test_socket.py (hunk 9 above)
Comment (caavery) [reply Lib/test/test_socket.py]: No problem.

File: aclocal.m4 (hunk 2 above)
Comment (caavery) [reply aclocal.m4]: I rebased my code to remove my patch. Is rebasing and pushing an acceptable way to make changes? Or do I need to add a new commit each time?

Thanks

Comment (caavery) [reply aclocal.m4]: Sorry to remove the aclocal.m4 from my patch.

File: Doc/library/socket.rst (hunk 8 above)
Comment (caavery) [reply Doc/library/socket.rst]: Well I thought it would be better informationally to have it in but if it is not acceptable I can remove it.

File: Lib/test/test_socket.py (hunk 3 above)
Comment (caavery) [reply Lib/test/test_socket.py]: is get_cid acceptable?

Comment (caavery) [reply Lib/test/test_socket.py]: Sorry wrong comment.

File: Lib/test/test_socket.py (hunk 9 above)
Comment (caavery) [reply Lib/test/test_socket.py]: is get_cid acceptable?

File: Lib/test/test_socket.py (hunk 3 above)
Comment (caavery) [reply Lib/test/test_socket.py]: I don't understand this syntax. It looks like return the cid if it is not None or do nothing and continue. getCid will return an integer or None. I want to translate that into return True or False in  _have_socket_vsock().

Comment (musically-ut) [reply Lib/test/test_socket.py]: @caavery read @berkerpeksag's suggestion as a contraction of the following:

```python
    foo = getCid is not None
    return foo
```
`foo` will contain a boolean value which will be returned.

Comment (caavery) [reply Lib/test/test_socket.py]: Thanks!

Comment (berkerpeksag) [reply Lib/test/test_socket.py]: Yeah, that's what I was trying to say. Sorry for being unclear!

And +1 for renaming it to `get_cid`, thanks!

File: Modules/socketmodule.c (hunk 1 above)
Comment (caavery) [reply Modules/socketmodule.c:1232]: I've restored the original aclocal.m4 and fixed the issues you pointed out. 

Thank you for your review,

Cathy

**Comment by berkerpeksag** [comment]
Looks like your branch contains additional commits and unfortunately it makes code review harder. Could you rebase your branch against current master? You might find https://devguide.python.org/gitbootcamp/#syncing-with-upstream helpful.

**Comment by caavery** [comment]
@berkerpeksag 

So you want me to sync my github branch bpo-27584 with the upstream cpython branch? Is that correct? I'm just a little unclear as to what you are asking.

Cathy

**Comment by caavery** [comment]
@berkerpeksag 

Or do you want me to rebase both my master ( fork of upstream ) and my bpo-27584 branch?

Thanks again

**Comment by berkerpeksag** [comment]
> So you want me to sync my github branch bpo-27584 with the upstream cpython branch? Is that correct? 

Correct. Your branch bpo-27584 now contains 34 commits and 30 of them are unrelated to the vSockets feature as you can see at https://github.com/python/cpython/pull/2489/files (the reason is probably [this merge commit](https://github.com/python/cpython/pull/2489/commits/4bc606af2514ab87ebba909f70994798243ef40a))

**Comment by caavery** [comment]
@berkerpeksag 

Sorry I seem to have an issue. 
git fetch upstream
git rebase upstream/master

Python Dev~/caavery.github/cpython# git push origin bpo-27584
To git@github.com:caavery/cpython.git
 ! [rejected]        bpo-27584 -> bpo-27584 (non-fast-forward)
error: failed to push some refs to 'git@github.com:caavery/cpython.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Merge the remote changes (e.g. 'git pull')
hint: before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

**Comment by caavery** [comment]
@berkerpeksag

Sorry this is my first time with github and working with remote branches. I just forced the push into my branch bpo-27584. My branch says that 'This branch is 4 commits ahead of python:master. ' which is what I think you wanted.

**Comment by berkerpeksag** [comment]
> My branch says that 'This branch is 4 commits ahead of python:master. ' which is what I think you wanted.

Looks great, thanks! I will try to take another look at the patch while we wait for Kushal's review.

**Comment by caavery** [comment]
@kushaldas 

Is there anything else I should be doing? The patches have passed all tests so far and I have addressed all the issues raised.

Thanks,

Cathy

File: Doc/library/socket.rst [hunk 11]
```diff
@@ -149,6 +149,14 @@ created.  Socket addresses are represented as follows:
 
   .. versionadded:: 3.6
 
+- :const:`AF_VSOCK` allows communication between virtual machines and
+  their hosts. The sockets are represented as a ``(CID,port)`` tuple
```
Comment (ilevkivskyi) [inline Doc/library/socket.rst]: Shouldn't there be space after coma? ``(CID, port)``.

Comment (caavery) [reply Doc/library/socket.rst]: Yes that got accidentally deleted.

File: Doc/library/socket.rst [hunk 12]
```diff
@@ -149,6 +149,14 @@ created.  Socket addresses are represented as follows:
 
   .. versionadded:: 3.6
 
+- :const:`AF_VSOCK` allows communication between virtual machines and
+  their hosts. The sockets are represented as a ``(CID, port)`` tuple
+  where the context ID or CID and port are integers.
+
+  Availability: Linux >= 4.8. QEMU >= 2.8.
```
Comment (rmsr) [inline Doc/library/socket.rst]: Given that vsockets originated with VMWare, add a note that they are available in that environment also?

File: Lib/test/test_socket.py [hunk 13]
```diff
@@ -44,6 +46,22 @@
 except ImportError:
     _socket = None
 
+def get_cid():
+    if fcntl is None:
+        return None
+    if not os.path.exists("/dev/vsock"):
```
Comment (tiran) [inline Lib/test/test_socket.py]: Please rewrite this block with proper context manager. You can skip the os.path.exists check, too.

```
try:
    with open(...) as f:
        r = fcntl.ioctl(f, socket.IOCTL_VM_SOCKETS_GET_LOCAL_CID, "    ")
except OSError:
    return None
else:
    return struct.unpack("I", r)[0]
```

File: Modules/socketmodule.c [hunk 14]
```diff
@@ -2608,20 +2665,39 @@ sock_getsockopt(PySocketSockObject *s, PyObject *args)
...
     if (!PyArg_ParseTuple(args, "ii|i:getsockopt",
                           &level, &optname, &buflen))
         return NULL;
 
     if (buflen == 0) {
-        int flag = 0;
-        socklen_t flagsize = sizeof flag;
+#ifdef AF_VSOCK
+        if (s->sock_family == AF_VSOCK) {
+            uint64_t vflag = 0;
```
Comment (tiran) [inline Modules/socketmodule.c]: I've been starring at this block for a minute until I realized that it handles ``uint64_t`` instead of ``int``. Please add a comment that this block that explains the difference.

File: Modules/socketmodule.c [hunk 15]
```diff
@@ -2608,20 +2665,39 @@ sock_getsockopt(PySocketSockObject *s, PyObject *args)
...
+            flagsize = sizeof vflag;
+            res = getsockopt(s->sock_fd, level, optname,
+                         (void *)&vflag, &flagsize);
+            if (res < 0)
+                return s->errorhandler();
+            return PyLong_FromLong(vflag);
```
Comment (tiran) [inline Modules/socketmodule.c]: It's an ``uint64_t``, you have to use ``PyLong_FromUnsignedLong()``.

File: Modules/socketmodule.h [hunk 16]
```diff
@@ -107,6 +107,15 @@ typedef int socklen_t;
 #define SOL_ALG 279
 #endif
 
+#ifdef HAVE_LINUX_VM_SOCKETS_H
+# include <linux/vm_sockets.h>
+#  ifndef AF_VSOCK
+#    define AF_VSOCK 40
```
Comment (tiran) [inline Modules/socketmodule.h]: Why is this necessary?

**Review by tiran — CHANGES_REQUESTED** [review]
I added some comments.

You also have to add the vsock struct to ``typedef union sock_addr`` in ``socketmodule.h``.

Please use ``blurb`` to add a news entry, too.

File: Doc/library/socket.rst (hunk 12 above)
Comment (caavery) [reply Doc/library/socket.rst]: This is a bit more complicated. The availability in a particular linux kernel depends on the implementation. 
virtio-vsock is Linux >= 4.8. QEMU >= 2.8.
vmci_transport is Linux >= 3.9
Should I just change it to when it first appeared which is v3.9?

File: Lib/test/test_socket.py (hunk 13 above)
Comment (caavery) [inline Lib/test/test_socket.py]: I seem to be having blurb issues:

**Comment by caavery** [comment]
I didn't expect the Spanish Inquisition!

File: Modules/socketmodule.h [hunk 17]
```diff
@@ -107,6 +107,12 @@ typedef int socklen_t;
 #define SOL_ALG 279
 #endif
 
+#ifdef HAVE_LINUX_VM_SOCKETS_H
+# include <linux/vm_sockets.h>
+#else
+# undef AF_VSOCK
```
Comment (tiran) [inline Modules/socketmodule.h:113]: I don't think ``#undef AF_VSOCK`` is necessary. Please remove it.

File: Modules/socketmodule.h [hunk 18]
```diff
@@ -193,6 +199,9 @@ typedef union sock_addr {
 #ifdef HAVE_SOCKADDR_ALG
     struct sockaddr_alg alg;
 #endif
+#ifdef HAVE_LINUX_VM_SOCKETS_H
```
Comment (tiran) [inline Modules/socketmodule.h]: Please be consistent and use ``#ifdef AF_VSOCK`` in all places.

**Review by tiran — CHANGES_REQUESTED** [review]
You are almost there!

File: Doc/library/socket.rst (hunk 12 above)
Comment (rmsr) [reply Doc/library/socket.rst]: I meant availability when running under the VMWare hypervisor. As for the kernel, to keep it simple, whichever version supplies full functionality with both qemu and vmware.

**Comment by caavery** [comment]
OK there seems to be a problem with the build now. I'll address it next week. Thanks.

File: Modules/socketmodule.h (hunk 17 above)
Comment (caavery) [reply Modules/socketmodule.h:113]: It needs to be there. In the case that vm_sockets.h is not available in include/linux which is the case on the last build bot run. Not all of the bots have the newer kernel.That way the code dependent on vm_sockets.h (AF_VSOCK) will not try to compile.

checking for linux/vm_sockets.h... no

This is also the way its done it at the top of socketmodule.h with AF_NETLINK

I also noticed AF_CAN was not adding the #undef AF_CAN in socketmodule.h so I renamed linux/can.h, ran configure and make and sure enough any code conditionally compiling under #ifdef AF_CAN in socketmodule.c failed.

**Comment by tiran** [comment]
You were right, either ``#undef AF_VSOCK`` is required or you have to replace all ``#ifdef AF_VSOCK`` with ``#ifdef HAVE_LINUX_VM_SOCKETS_H``. The address family and the VM socket struct are defined in different header files. ``AF_VSOCK`` can be present w/o ``linux/vm_sockets.h``.

**Comment by caavery** [comment]
I didn't expect the Spanish Inquisition!

**Review by tiran — APPROVED** [review]
LGTM, thanks!

**Comment by vstinner** [comment]
FYI I wrote PR gh-119463 to skip the tests if get_cid() returns VMADDR_CID_ANY. Apparently, the Linux kernel 6.9 changed `/dev/vsock` device.
